File: src/voyage_geo/stages/analysis/analyzers/positioning.py
"""Positioning analyzer — how AI models describe/position the brand."""

from __future__ import annotations

from collections import Counter

from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

from voyage_geo.types.analysis import PositionAttribute, PositioningScore
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.result import QueryResult
from voyage_geo.utils.text import contains_brand, extract_sentences

vader = SentimentIntensityAnalyzer()

POSITION_KEYWORDS = [
    "leader", "popular", "best", "top", "innovative", "affordable", "reliable",
    "powerful", "simple", "enterprise", "scalable", "trusted", "fast", "secure",
    "flexible", "comprehensive", "user-friendly", "modern", "mature", "growing",
    "niche", "expensive", "complex", "limited", "outdated", "basic",
]


class PositioningAnalyzer:
    name = "positioning"

    def analyze(self, results: list[QueryResult], profile: BrandProfile) -> PositioningScore:
        valid = [r for r in results if not r.error and r.response]

        attr_counter: Counter[str] = Counter()
        attr_sentiments: dict[str, list[float]] = {}
        by_provider: dict[str, str] = {}

        for r in valid:
            sentences = [s for s in extract_sentences(r.response) if contains_brand(s, profile.name)]
            provider_attrs: Counter[str] = Counter()

            for sentence in sentences:
                lower = sentence.lower()
                for kw in POSITION_KEYWORDS:
                    if kw in lower:
                        attr_counter[kw] += 1
                        provider_attrs[kw] += 1
                        vs = vader.polarity_scores(sentence)
                        attr_sentiments.setdefault(kw, []).append(vs["compound"])

            if provider_attrs and r.provider not in by_provider:
                by_provider[r.provider] = provider_attrs.most_common(1)[0][0]

        attributes = []
        for attr, freq in attr_counter.most_common(10):
            import statistics
            avg_sent = statistics.mean(attr_sentiments.get(attr, [0]))
            attributes.append(PositionAttribute(attribute=attr, frequency=freq, sentiment=round(avg_sent, 3)))

        primary = attr_counter.most_common(1)[0][0] if attr_counter else "unknown"

        return PositioningScore(
            primary_position=primary,
            attributes=attributes,
            by_provider=by_provider,
        )

File: src/stages/analysis/analyzers/positioning.ts
import type { Analyzer } from './types.js';
import type { QueryResult } from '../../../types/result.js';
import type { BrandProfile } from '../../../types/brand.js';
import type { PositioningScore, PositionAttribute } from '../../../types/analysis.js';
import { containsBrand, extractSentences } from '../../../utils/text.js';
import { groupBy } from '../statistics.js';
import Sentiment from 'sentiment';

const sentimentAnalyzer = new Sentiment();

const POSITION_KEYWORDS = [
  'leader', 'best', 'top', 'popular', 'powerful', 'innovative', 'reliable',
  'affordable', 'premium', 'enterprise', 'simple', 'easy', 'fast', 'secure',
  'scalable', 'flexible', 'modern', 'comprehensive', 'lightweight', 'robust',
  'free', 'open-source', 'collaborative', 'intuitive', 'customizable',
];

export class PositioningAnalyzer implements Analyzer {
  name = 'positioning';

  analyze(results: QueryResult[], profile: BrandProfile): PositioningScore {
    const validResults = results.filter((r) => !r.error && r.response);

    // Find attributes associated with the brand
    const attributeCounts = new Map<string, { count: number; sentiments: number[] }>();

    for (const result of validResults) {
      const sentences = extractSentences(result.response)
        .filter((s) => containsBrand(s, profile.name));

      for (const sentence of sentences) {
        const lower = sentence.toLowerCase();
        for (const keyword of POSITION_KEYWORDS) {
          if (lower.includes(keyword)) {
            const existing = attributeCounts.get(keyword) || { count: 0, sentiments: [] };
            existing.count++;
            existing.sentiments.push(sentimentAnalyzer.analyze(sentence).comparative);
            attributeCounts.set(keyword, existing);
          }
        }
      }
    }

    // Sort by frequency
    const attributes: PositionAttribute[] = Array.from(attributeCounts.entries())
      .map(([attribute, data]) => ({
        attribute,
        frequency: data.count,
        sentiment: data.sentiments.reduce((a, b) => a + b, 0) / data.sentiments.length,
      }))
      .sort((a, b) => b.frequency - a.frequency);

    const primaryPosition = attributes.length > 0
      ? attributes.slice(0, 3).map((a) => a.attribute).join(', ')
      : 'unpositioned';

    // By provider
    const byProvider: Record<string, string> = {};
    const providerGroups = groupBy(validResults, (r) => r.provider);
    for (const [provider, providerResults] of providerGroups) {
      const providerAttributes = new Map<string, number>();
      for (const result of providerResults) {
        const sentences = extractSentences(result.response)
          .filter((s) => containsBrand(s, profile.name));
        for (const sentence of sentences) {
          const lower = sentence.toLowerCase();
          for (const keyword of POSITION_KEYWORDS) {
            if (lower.includes(keyword)) {
              providerAttributes.set(keyword, (providerAttributes.get(keyword) || 0) + 1);
            }
          }
        }
      }
      const top = Array.from(providerAttributes.entries())
        .sort(([, a], [, b]) => b - a)
        .slice(0, 3)
        .map(([k]) => k)
        .join(', ');
      byProvider[provider] = top || 'unpositioned';
    }

    return {
      primaryPosition,
      attributes: attributes.slice(0, 10),
      byProvider,
    };
  }
}
